casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -303,47 +299,66 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
-      //TODO: compare replay deploy cost to given deploy cost
-      EvaluateResult(_, errors) = replayEvaluateResult
-      _                         <- Span[F].mark("before-replay-deploy-status")
-      cont <- DeployStatus.fromErrors(errors) match {
-               case int: InternalErrors =>
-                 (deploy.some, int: Failed).some.pure[F]
-               case replayStatus =>
-                 if (status.isFailed != replayStatus.isFailed)
-                   (deploy.some, ReplayStatusMismatch(replayStatus, status): Failed).some
-                     .pure[F]
-                 else if (errors.nonEmpty)
-                   runtime.replaySpace.revertToSoftCheckpoint(softCheckpoint) >> none[ReplayFailure]
-                     .pure[F]
-                 else {
-                   runtime.replaySpace
-                     .checkReplayData()
-                     .attempt
-                     .flatMap {
-                       case Right(_) => none[ReplayFailure].pure[F]
-                       case Left(ex: ReplayException) => {
-                         Log[F].error(s"Failed during processing of deploy: ${processedDeploy}") >>
-                           (none[DeployData], UnusedCommEvent(ex): Failed).some
-                             .pure[F]
-                       }
-                       case Left(ex) =>
-                         (none[DeployData], UserErrors(Vector(ex)): Failed).some
-                           .pure[F]
-                     }
-                 }
-             }
-    } yield cont
+      replayEvaluateResultEither <- evaluate(runtime.replayReducer, runtime.cost, runtime.errorLog)(
+                                     processedDeploy.deploy
+                                   ).attempt
+      failureOption <- replayEvaluateResultEither match {
+                        case Right(EvaluateResult(replayCost, replayUserErrors)) =>
+                          /* Regardless of success or failure, verify that deploy status' match. */
+                          if (isFailed != replayUserErrors.nonEmpty)
+                            runtime.replaySpace
+                              .revertToSoftCheckpoint(softCheckpoint)
+                              .as(
+                                ReplayFailure
+                                  .replayStatusMismatch(isFailed, replayUserErrors.nonEmpty)
+                                  .some
+                              )
+                          /* Error-throwing Rholang programs do not have deterministic evaluation costs
+                             and event logs, so they cannot be reliably compared. */
+                          else if (replayUserErrors.nonEmpty)
+                            runtime.replaySpace
+                              .revertToSoftCheckpoint(softCheckpoint)
+                              .as(none[ReplayFailure])
+                          /* Since there are no errors, verify evaluation costs and COMM events match. */
```

Comment:
Those are one of the most helpful comments I've ever seen in code, thank you! :)